doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,271 @@
+      of the API and bugs caused by misuse.
+ - [Reactor pattern](https://en.wikipedia.org/wiki/Reactor_pattern)
+ - [Condition Variable](https://en.wikipedia.org/wiki/Monitor_(synchronization)#Condition_variables_2)
+ - [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug)
+ 
+## Design
+
+### General
+The usage should be similar to the _WaitSet_ with a key difference - it should 
+be **event driven** and not a mixture of event and state driven, depending on
+which event is assigned, like in the _WaitSet_.
+
+It should have the following behavior:
+
+ - Whenever an event occurs the corresponding callback should be called **once**
+    as soon as possible.
+ - If an event occurs multiple times before the callback was called, the callback
+    should be called **once**.
+ - If an event occurs while the callback is being executed the callback should be 
+    called again **once**.
+
+The following use cases and behaviors should be implemented.
+
+ - The API and ReactAL should be robust. Since ReactAL is working concurrently a 
+    wrong usage could lead to race conditions, extremely hard to debug bug reports
+    (Heisenbugs) and can be frustrating to the user.
+    We list here features marked with [robust] which are only supported to 
+    increase this kind of the robustness.
+
+ - concurrently: assigning a callback at any time from anywhere. This means in particular
+    - Attaching a callback from within a callback.
+    - [robust] Updating a callback from within the same callback 
+      ```cpp
+      void onSampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived2);
+      }
+
+      myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+      ```
+ - One can assign at most one callback to a specific event of an object. The event is usually defined 
+    with an enum by the developer. One example is `SubscriberEvent::HAS_SAMPLE_RECEIVED`.
+    - Attaching a callback to an event where a callback has been already assigned overrides
+      the existing callback with the new one.
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
+                 and not until the callback is finished. After the successful removal the 
+                 callback will continue and will not be called again.
+    - Calling detach means that after that call the callback is no longer assigned
```

Comment:
This could be phrased a bit simpler
```
The callback is no longer assigned to the event after calling `detach`. The method will succeed even if it was not assigned in the first place.
```